clash-lib/src/Clash/Netlist/Types.hs

```diff
@@ -421,6 +449,70 @@ instance Binary TemplateFunction where
   get = (\is -> TemplateFunction is err err) <$> get
     where err = const $ error "TemplateFunction functions can't be preserved by serialisation"
 
+-- | Netlist-level identifier
+data NetlistId
+  = NetlistId Identifier Type
+  -- ^ Identifier generated in the NetlistMonad, always derived from another
+  -- 'NetlistId'
+  | CoreId Id
+  -- ^ An original Core identifier
+  | MultiId [Id]
+  -- ^ A split identifier (into several sub-identifiers), needed to assign
+  -- expressions of types that have to be split apart (e.g. tuples of Files)
+  deriving Show
+
+-- | Eliminator for 'NetlistId'
+netlistId
+  :: (Identifier -> r)
+  -- ^ Eliminator for Identifiers generated in the NetlistMonad
+  -> (Id -> r)
+  -- ^ Eliminator for original Core Identifiers
+  -> NetlistId
+  -> [r]
+netlistId f g = \case
+  NetlistId i _ -> [f i]
+  CoreId i      -> [g i]
+  MultiId is    -> map g is
+
+-- | Eliminator for 'NetlistId', fails on 'MultiId'
+netlistId1
+  :: HasCallStack
+  => (Identifier -> r)
+  -- ^ Eliminator for Identifiers generated in the NetlistMonad
+  -> (Id -> r)
+  -- ^ Eliminator for original Core Identifiers
+  -> NetlistId
+  -> r
+netlistId1 f g = \case
+  NetlistId i _ -> f i
+  CoreId i      -> g i
+  m             -> error ("netlistId1 MultiId: " ++ show m)
+
+-- | Return the type(s) of a 'NetListId', returns multiple types when given a
+-- 'MultiId'
+netlistTypes
+  :: NetlistId
+  -> [Type]
+netlistTypes = \case
+  NetlistId _ t -> [t]
+  CoreId i      -> [varType i]
+  MultiId is    -> map varType is
```

Comment:
Minor nitpick, but I don't think we should align things like this. I don't think it helps readability and it makes an refactors (added alternatives / renames) tedious to do correctly.

Reply:
Changed